File: sdks/python/src/rcm/mcp.py
"""MCP server configuration for the RCM gRPC SDK.

Mirrors the .rcm language `mcp` block.
"""

from dataclasses import dataclass, field
from typing import Optional


@dataclass
class McpServer:
    label: str
    transport: "McpServer.Stdio | McpServer.Http | McpServer.Sse"

    @staticmethod
    def Env(name: str):
        return McpServer.Value(env=name)

    @dataclass
    class Value:
        literal: Optional[str] = None
        env: Optional[str] = None

    @dataclass
    class Stdio:
        command: str
        args: list[str] = field(default_factory=list)
        env: dict[str, "McpServer.Value"] = field(default_factory=dict)
        cwd: Optional[str] = None

    @dataclass
    class Http:
        url: str
        headers: dict[str, "McpServer.Value"] = field(default_factory=dict)

    @dataclass
    class Sse:
        url: str
        headers: dict[str, "McpServer.Value"] = field(default_factory=dict)

    def _to_proto(self):
        from ._pb2 import (
            McpHttpSpec,
            McpServerSpec,
            McpSseSpec,
            McpStdioSpec,
            McpTransportSpec,
            McpValueSpec,
        )

        def _value_proto(value: "McpServer.Value"):
            if value.env is not None:
                return McpValueSpec(env=value.env)
            if value.literal is not None:
                return McpValueSpec(literal=value.literal)
            return McpValueSpec()

        spec = McpServerSpec(label=self.label)
        transport = self.transport

        if isinstance(transport, McpServer.Stdio):
            spec.transport.CopyFrom(
                McpTransportSpec(
                    stdio=McpStdioSpec(
                        command=transport.command,
                        args=transport.args,
                        env={k: _value_proto(v) for k, v in transport.env.items()},
                        cwd=transport.cwd,
                    )
                )
            )
        elif isinstance(transport, McpServer.Http):
            spec.transport.CopyFrom(
                McpTransportSpec(
                    http=McpHttpSpec(
                        url=transport.url,
                        headers={
                            k: _value_proto(v) for k, v in transport.headers.items()
                        },
                    )
                )
            )
        elif isinstance(transport, McpServer.Sse):
            spec.transport.CopyFrom(
                McpTransportSpec(
                    sse=McpSseSpec(
                        url=transport.url,
                        headers={
                            k: _value_proto(v) for k, v in transport.headers.items()
                        },
                    )
                )
            )

        return spec

File: crates/server/src/mcp.rs
use accelerator::mcp::{McpServerConfig, McpTransportConfig};
use std::collections::HashMap;
use tonic::Status;

pub fn mcp_config_from_spec(spec: &crate::rcm::McpServerSpec) -> Result<McpServerConfig, Status> {
    let transport = spec
        .transport
        .as_ref()
        .ok_or(Status::invalid_argument("mcp requires transport"))?;
    use crate::rcm::mcp_transport_spec::Kind;
    let transport = match &transport.kind {
        Some(Kind::Stdio(stdio)) => McpTransportConfig::Stdio {
            command: stdio.command.clone(),
            args: stdio.args.clone(),
            env: mcp_env_from_spec(&stdio.env)?,
            cwd: stdio.cwd.clone().map(std::path::PathBuf::from),
        },
        Some(Kind::Http(http)) => McpTransportConfig::Http {
            url: http.url.clone(),
            headers: mcp_headers_from_spec(&http.headers)?,
        },
        Some(Kind::Sse(sse)) => McpTransportConfig::Sse {
            url: sse.url.clone(),
            headers: mcp_headers_from_spec(&sse.headers)?,
        },
        None => return Err(Status::invalid_argument("mcp transport kind required")),
    };
    Ok(McpServerConfig {
        label: spec.label.clone(),
        transport,
    })
}

fn mcp_env_from_spec(
    values: &HashMap<String, crate::rcm::McpValueSpec>,
) -> Result<HashMap<String, String>, Status> {
    values
        .iter()
        .map(|(name, value)| Ok((name.clone(), mcp_value_text(value)?)))
        .collect()
}

fn mcp_headers_from_spec(
    values: &HashMap<String, crate::rcm::McpValueSpec>,
) -> Result<Vec<(String, String)>, Status> {
    values
        .iter()
        .map(|(name, value)| Ok((name.clone(), mcp_value_text(value)?)))
        .collect()
}

fn mcp_value_text(value: &crate::rcm::McpValueSpec) -> Result<String, Status> {
    use crate::rcm::mcp_value_spec::Source;
    match &value.source {
        Some(Source::Literal(value)) => Ok(value.clone()),
        Some(Source::Env(name)) => std::env::var(name).map_err(|error| {
            Status::invalid_argument(format!("env var '{name}' not set: {error}"))
        }),
        None => Ok(String::new()),
    }
}
